File: DSA-lab/modules/02-arrays-strings/projects/02-medium/catalog_demo.py
"""Run one representative example for every pattern in the catalog."""

from pattern_catalog import (
    PrefixSum,
    are_anagrams,
    compress_runs_in_place,
    longest_unique_substring,
    merge_sorted_in_place,
    move_zeros_to_end,
    product_except_self,
    remove_duplicates_sorted,
    rotate_right_in_place,
    spiral_order,
)


def main() -> None:
    """Execute and print all ten catalog examples."""
    rotated = [1, 2, 3, 4, 5, 6, 7]
    rotate_right_in_place(rotated, 3)

    merged = [1, 4, 7, None, None, None]
    merge_sorted_in_place(merged, 3, [2, 3, 8])

    prefix = PrefixSum.from_values([4, -1, 7, 3, 2])

    unique_values = [1, 1, 2, 2, 2, 5]
    remove_duplicates_sorted(unique_values)

    moved_zeros = [0, 3, 0, 1, 0, 8]
    move_zeros_to_end(moved_zeros)

    compressed = list("aaabcccccccccc")
    compress_runs_in_place(compressed)

    examples = [
        ("Rotation", rotated),
        ("Merge", merged),
        ("Prefix sum [1:4]", prefix.range_sum(1, 4)),
        ("Anagram", are_anagrams("listen", "silent")),
        ("Deduplizierung", unique_values),
        ("Zero-Move", moved_zeros),
        ("Product Except Self", product_except_self([1, 2, 3, 4])),
        ("Längstes eindeutiges Teilstück", longest_unique_substring("pwwkew")),
        ("Spirale", spiral_order([[1, 2, 3], [4, 5, 6], [7, 8, 9]])),
        ("Kompression", "".join(compressed)),
    ]

    for name, result in examples:
        print(f"{name:34} {result}")


if __name__ == "__main__":
    main()

File: DSA-lab-java/modules/02-arrays-strings/projects/02-medium/src/main/java/lab/dsa/module02/medium/CatalogDemo.java
package lab.dsa.module02.medium;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/** Runs one representative example for every catalog pattern. */
public final class CatalogDemo {
    private CatalogDemo() {
    }

    public static void main(String[] arguments) {
        List<Integer> rotated = new ArrayList<>(List.of(1, 2, 3, 4, 5, 6, 7));
        PatternCatalog.rotateRightInPlace(rotated, 3);
        List<Integer> merged = new ArrayList<>(Arrays.asList(1, 4, 7, null, null, null));
        PatternCatalog.mergeSortedInPlace(merged, 3, List.of(2, 3, 8));
        PatternCatalog.PrefixSum prefix = PatternCatalog.PrefixSum.fromValues(List.of(4, -1, 7, 3, 2));
        List<Integer> unique = new ArrayList<>(List.of(1, 1, 2, 2, 2, 5));
        PatternCatalog.removeDuplicatesSorted(unique);
        List<Number> zeros = new ArrayList<>(List.of(0, 3, 0, 1, 0, 8));
        PatternCatalog.moveZerosToEnd(zeros);
        List<Character> compressed = new ArrayList<>();
        for (char value : "aaabcccccccccc".toCharArray()) compressed.add(value);
        PatternCatalog.compressRunsInPlace(compressed);
        System.out.println("Rotation                         " + rotated);
        System.out.println("Merge                            " + merged);
        System.out.println("Prefix sum [1:4]                 " + prefix.rangeSum(1, 4));
        System.out.println("Anagram                          " + PatternCatalog.areAnagrams("listen", "silent"));
        System.out.println("Deduplication                    " + unique);
        System.out.println("Move zeros                       " + zeros);
        System.out.println("Product except self              " + Arrays.toString(PatternCatalog.productExceptSelf(new long[]{1, 2, 3, 4})));
        System.out.println("Longest unique substring         " + PatternCatalog.longestUniqueSubstring("pwwkew"));
        System.out.println("Spiral                           " + PatternCatalog.spiralOrder(List.of(List.of(1, 2, 3), List.of(4, 5, 6), List.of(7, 8, 9))));
        StringBuilder encoded = new StringBuilder(); compressed.forEach(encoded::append);
        System.out.println("Compression                      " + encoded);
    }
}
